infra/src/api/io_services_cms/v1/createservice_policy/policy.xml

```diff
@@ -6,6 +6,30 @@
             (because every subscription creation modifies the arm resource).
             In order to mitigate so, we limit concurrency ourself.-->
         <limit-concurrency key="constant-key-all-users" max-count="10" />
+        <choose>
+            <when condition="@(context.Request.Body != null)">
+                <set-body>
+@{
+  var stringBody = context.Request.Body.As<string>();
+  try {
+    var jsonBody = JObject.Parse(stringBody);
+    try {
+      if (context.Request.Headers.ContainsKey("x-user-groups-selc")) {
+        jsonBody["data"]["groupId"] = context.Request.Headers["x-user-groups-selc"][0];
+      } else {
+        jsonBody["data"]["groupId"].Parent.Remove();
+      }
+      return jsonBody.ToString();
+    } catch (Exception) {
+      return jsonBody.ToString();
```

Comment:
Do you think it could be useful to use the `trace` policy to track when the exception is raised for troubleshooting purposes? https://learn.microsoft.com/en-us/azure/api-management/trace-policy

Reply:
Actually these policy are only useful only to ensure "robustness" of these two operations (createService and editService), because this new field (groupId) will not be added to "external" OpenAPI, so clients using these APIs should not even be aware of the existence of this field.
Actually, these exceptions could be raised only in these two cases:
- when the request body is not a json (`JObject.Parse(stringBody)` fails)
- when the request body has not a `data` property (`jsonBody["data"]["groupId"]` fails)
Both cases are managed by the application/backend request handler (and, in case of failure, also traces via Application Insight).

So, I think there is no need to add the trace here.

PS: an exception could be raised also in cases when caller do not insert groupId property (`jsonBody["data"]["groupId"].Parent.Remove()` fails), but, in this case, the absence of this field is exactly what we want, so it is OK to leave the request as it came to us.

Reply:
Basically, all request checking/validation is done by the application/backend (and also traced via App Insight), this policy is only to ensure that from this channel (APIM) the groupId field is not manipulated by the caller.